src/osd/OSD.cc

```diff
@@ -6474,26 +6494,40 @@ bool OSD::ms_handle_refused(Connection *con)
   dout(2) << "ms_handle_refused con " << con << " session " << session.get() << dendl;
   if (!session)
     return false;
+
   int type = con->get_peer_type();
   // handle only OSD failures here
   if (monc && (type == CEPH_ENTITY_TYPE_OSD)) {
+    map_lock.lock_shared();
+    std::lock_guard l(mon_report_lock);
+    std::lock_guard l2(heartbeat_lock);
+
     OSDMapRef osdmap = get_osdmap();
     if (osdmap) {
       int id = osdmap->identify_osd_on_all_channels(con->get_peer_addr());
       if (id >= 0 && osdmap->is_up(id)) {
 	// I'm cheating mon heartbeat grace logic, because we know it's not going
 	// to respawn alone. +1 so we won't hit any boundary case.
-	monc->send_mon_message(
-	  new MOSDFailure(
-	    monc->get_fsid(),
-	    id,
-	    osdmap->get_addrs(id),
-	    cct->_conf->osd_heartbeat_grace + 1,
-	    osdmap->get_epoch(),
-	    MOSDFailure::FLAG_IMMEDIATE | MOSDFailure::FLAG_FAILED
-	    ));
+	if (!failure_immediate_pending.count(id)) {
```

Comment:
Why not just reuse failure_pending here

Reply:
When doing this optimization, the main reason is not to change the original osd-osd heartbeat timeout logic. It is necessary to think deeply about whether reuse the failure_pending can introduce problems.

Reply:
Yeah, that needs to be thought through either way.  This seems to be a special case of what failure_pending already does and I'd rather generalize that than introduce a new thing.

Reply:
Another problem is that MOSDFailure(FLAG_FAILED) in failure_pending is now added by osd-osd heartbeat timeout, and its behavior in mon is different from MOSDFailure(FLAG_IMMEDIATE|FLAG_FAILED),
MOSDFailure(FLAG_FAILED) Timeout failed: to wait for mon_osd_min_down_reporters
MOSDFailure(FLAG_IMMEDIATE|FLAG_FAILED) fail immediately : no waiting

Reply:
```
bool OSDMonitor::prepare_failure(MonOpRequestRef op)
{
    auto m = op->get_req<MOSDFailure>();
    if (m->if_osd_failed()) {
        // calculate failure time
        utime_t now = ceph_clock_now();
        utime_t failed_since = m->get_recv_stamp() - utime_t(m->failed_for, 0);

        // add a report
        if (m->is_immediate()) {   ///MOSDFailure(FLAG_IMMEDIATE|FLAG_FAILED) 
            mon->clog->debug() << "osd." << m->get_target_osd() << " reported immediately failed by "
                               << m->get_orig_source();
            force_failure(target_osd, reporter);  ///MOSDFailure(FLAG_IMMEDIATE|FLAG_FAILED) 
            return true;
        }
        mon->clog->debug() << "osd." << m->get_target_osd() << " reported failed by " << m->get_orig_source();

        failure_info_t &fi = failure_info[target_osd];
        fi.add_report(reporter, failed_since, op);
        return check_failure(now, target_osd, fi);  ///MOSDFailure(FLAG_FAILED) 
    }
}
```

Reply:
Replace the std::pair<utime_t, entity_addrvec_t> with something like

```
struct {
  enum class failure_type_t {
    IMMEDIATE,
    FAILED
  } type;
  utime_t failure_time; // only valid if type == FAILED
  entity_addrvec_t addr
};
```

We don't want to send FLAG_IMMEDIATE MOSDFailure messages after we've sent a FLAG_FAILED message anyway, right?

Reply:
osd_fast_fail_on_connection_refused 
If this option is enabled, crashed OSDs are maked down immediately by connected peers and MONs (assuming that the crashed OSD host survives). Disable it to restore old behavior, at the expense of possible long I/O stalls when OSDs crash in the middle of I/O operations.

The difference between MOSDFailure(FLAG_IMMEDIATE|FLAG_FAILED) and MOSDFailure(FLAG_FAILED) is whether mon should wait for mon_osd_min_down_reporters to mark the target osd as down.

I think even if MOSDFailure(FLAG_FAILED) has been sent, in this case, MOSDFailure(FLAG_IMMEDIATE|FLAG_FAILED) is sent again

Reply:
Override MOSDFailure(FLAG_FAILED) in failure_pending with MOSDFailure(FLAG_IMMEDIATE|FLAG_FAILED)

Reply:
> Replace the std::pair<utime_t, entity_addrvec_t> with something like
> 
> ```
> struct {
>   enum class failure_type_t {
>     IMMEDIATE,
>     FAILED
>   } type;
>   utime_t failure_time; // only valid if type == FAILED
>   entity_addrvec_t addr;
> };
> ```
> 
> We don't want to send FLAG_FAILURE MOSDFailure messages after we've sent a FLAG_IMMEDIATE message anyway, right?

right